### ISSUE TITLE
The state property of the payload is missing on the BlockActionPayload

### ISSUE THREAD

**Issue by lfbrandao** [slackapi/java-slack-sdk#635]
I'm using Bolt in Kotlin and Spring Boot and I have a view with a few External Options Selects and a button. When I press the button I would like to be able to see what was selected on the selects.

The request payload sent from Slack includes a `state` property that has what I'm looking for (`selected_option`):

```
{
	"type": "block_actions",
	"state": {
		"values": {
			"project-selector-bcb56caa-38a3-4b95-b809-0eda418d5b51": {
				"checkin-selected-project": {
					"type": "external_select",
					"selected_option": {
						"text": {
							"type": "plain_text",
							"text": "some text",
							"emoji": false
						},
						"value": "some value"
					}
				}
			}
		},
		"response_url": "....",
		"actions": [{
			"action_id": "save",
			"block_id": "fdo",
			"text": {
				"type": "plain_text",
				"text": "Save",
				"emoji": true
			},
			"style": "primary",
			"type": "button",
			"action_ts": "..."
		}]
	}
}
``` 

but on my action handler I can't find the state:

```
app.blockAction("save") { req, ctx ->
    // req.payload.view is null
    // req.payload.view.state is a view property, but the view is null
    // req.payload.state.values is null
    ....
}
```

Am I looking at the wrong place or is this something that is not implemented yet?

### Reproducible in:

```bash
mvn dependency:tree | grep com.slack.api
gradle dependencies | grep com.slack.api
java -version
sw_vers && uname -v # or `ver`
```

#### The Slack SDK version

```
```bash
+--- com.slack.api:bolt:1.3.2
|    +--- com.slack.api:slack-api-model:1.3.2
|    +--- com.slack.api:slack-api-client:1.3.2
|    |    +--- com.slack.api:slack-api-model:1.3.2 (*)
|    \--- com.slack.api:slack-app-backend:1.3.2
|         +--- com.slack.api:slack-api-model:1.3.2 (*)
|         +--- com.slack.api:slack-api-client:1.3.2 (*)
+--- com.slack.api:slack-api-client:1.3.2 (*)
+--- com.slack.api:slack-api-model-kotlin-extension:1.3.2
|    +--- com.slack.api:slack-api-model:1.3.2 (*)
+--- com.slack.api:slack-api-client-kotlin-extension:1.3.2
|    +--- com.slack.api:slack-api-model-kotlin-extension:1.3.2 (*)
|    +--- com.slack.api:slack-api-client:1.3.2 (*)
+--- com.slack.api:bolt-servlet:1.3.2
|    +--- com.slack.api:slack-api-model:1.3.2 (*)
|    +--- com.slack.api:slack-api-client:1.3.2 (*)
|    +--- com.slack.api:slack-app-backend:1.3.2 (*)
|    \--- com.slack.api:bolt:1.3.2 (*)
+--- com.slack.api:bolt:1.3.2 (n)
+--- com.slack.api:slack-api-client:1.3.2 (n)
+--- com.slack.api:slack-api-model-kotlin-extension:1.3.2 (n)
+--- com.slack.api:slack-api-client-kotlin-extension:1.3.2 (n)
+--- com.slack.api:bolt-servlet:1.3.2 (n)
+--- com.slack.api:bolt:1.3.2
|    +--- com.slack.api:slack-api-model:1.3.2
|    +--- com.slack.api:slack-api-client:1.3.2
|    |    +--- com.slack.api:slack-api-model:1.3.2 (*)
|    \--- com.slack.api:slack-app-backend:1.3.2
|         +--- com.slack.api:slack-api-model:1.3.2 (*)
|         +--- com.slack.api:slack-api-client:1.3.2 (*)
+--- com.slack.api:slack-api-client:1.3.2 (*)
+--- com.slack.api:slack-api-model-kotlin-extension:1.3.2
|    +--- com.slack.api:slack-api-model:1.3.2 (*)
+--- com.slack.api:slack-api-client-kotlin-extension:1.3.2
|    +--- com.slack.api:slack-api-model-kotlin-extension:1.3.2 (*)
|    +--- com.slack.api:slack-api-client:1.3.2 (*)
+--- com.slack.api:bolt-servlet:1.3.2
|    +--- com.slack.api:slack-api-model:1.3.2 (*)
|    +--- com.slack.api:slack-api-client:1.3.2 (*)
|    +--- com.slack.api:slack-app-backend:1.3.2 (*)
|    \--- com.slack.api:bolt:1.3.2 (*)
+--- com.slack.api:bolt:1.3.2
|    +--- com.slack.api:slack-api-model:1.3.2
|    +--- com.slack.api:slack-api-client:1.3.2
|    |    +--- com.slack.api:slack-api-model:1.3.2 (*)
|    \--- com.slack.api:slack-app-backend:1.3.2
|         +--- com.slack.api:slack-api-model:1.3.2 (*)
|         +--- com.slack.api:slack-api-client:1.3.2 (*)
+--- com.slack.api:slack-api-client:1.3.2 (*)
+--- com.slack.api:slack-api-model-kotlin-extension:1.3.2
|    +--- com.slack.api:slack-api-model:1.3.2 (*)
+--- com.slack.api:slack-api-client-kotlin-extension:1.3.2
|    +--- com.slack.api:slack-api-model-kotlin-extension:1.3.2 (*)
|    +--- com.slack.api:slack-api-client:1.3.2 (*)
+--- com.slack.api:bolt-servlet:1.3.2
|    +--- com.slack.api:slack-api-model:1.3.2 (*)
|    +--- com.slack.api:slack-api-client:1.3.2 (*)
|    +--- com.slack.api:slack-app-backend:1.3.2 (*)
|    \--- com.slack.api:bolt:1.3.2 (*)
+--- com.slack.api:bolt:1.3.2
|    +--- com.slack.api:slack-api-model:1.3.2
|    +--- com.slack.api:slack-api-client:1.3.2
|    |    +--- com.slack.api:slack-api-model:1.3.2 (*)
|    \--- com.slack.api:slack-app-backend:1.3.2
|         +--- com.slack.api:slack-api-model:1.3.2 (*)
|         +--- com.slack.api:slack-api-client:1.3.2 (*)
+--- com.slack.api:slack-api-client:1.3.2 (*)
+--- com.slack.api:slack-api-model-kotlin-extension:1.3.2
|    +--- com.slack.api:slack-api-model:1.3.2 (*)
+--- com.slack.api:slack-api-client-kotlin-extension:1.3.2
|    +--- com.slack.api:slack-api-model-kotlin-extension:1.3.2 (*)
|    +--- com.slack.api:slack-api-client:1.3.2 (*)
+--- com.slack.api:bolt-servlet:1.3.2
|    +--- com.slack.api:slack-api-model:1.3.2 (*)
|    +--- com.slack.api:slack-api-client:1.3.2 (*)
|    +--- com.slack.api:slack-app-backend:1.3.2 (*)
|    \--- com.slack.api:bolt:1.3.2 (*)
+--- com.slack.api:bolt:1.3.2
|    +--- com.slack.api:slack-api-model:1.3.2
|    +--- com.slack.api:slack-api-client:1.3.2
|    |    +--- com.slack.api:slack-api-model:1.3.2 (*)
|    \--- com.slack.api:slack-app-backend:1.3.2
|         +--- com.slack.api:slack-api-model:1.3.2 (*)
|         +--- com.slack.api:slack-api-client:1.3.2 (*)
+--- com.slack.api:slack-api-client:1.3.2 (*)
+--- com.slack.api:slack-api-model-kotlin-extension:1.3.2
|    +--- com.slack.api:slack-api-model:1.3.2 (*)
+--- com.slack.api:slack-api-client-kotlin-extension:1.3.2
|    +--- com.slack.api:slack-api-model-kotlin-extension:1.3.2 (*)
|    +--- com.slack.api:slack-api-client:1.3.2 (*)
+--- com.slack.api:bolt-servlet:1.3.2
|    +--- com.slack.api:slack-api-model:1.3.2 (*)
|    +--- com.slack.api:slack-api-client:1.3.2 (*)
|    +--- com.slack.api:slack-app-backend:1.3.2 (*)
|    \--- com.slack.api:bolt:1.3.2 (*)
+--- com.slack.api:bolt:1.3.2
|    +--- com.slack.api:slack-api-model:1.3.2
|    +--- com.slack.api:slack-api-client:1.3.2
|    |    +--- com.slack.api:slack-api-model:1.3.2 (*)
|    \--- com.slack.api:slack-app-backend:1.3.2
|         +--- com.slack.api:slack-api-model:1.3.2 (*)
|         +--- com.slack.api:slack-api-client:1.3.2 (*)
+--- com.slack.api:slack-api-client:1.3.2 (*)
+--- com.slack.api:slack-api-model-kotlin-extension:1.3.2
|    +--- com.slack.api:slack-api-model:1.3.2 (*)
+--- com.slack.api:slack-api-client-kotlin-extension:1.3.2
|    +--- com.slack.api:slack-api-model-kotlin-extension:1.3.2 (*)
|    +--- com.slack.api:slack-api-client:1.3.2 (*)
+--- com.slack.api:bolt-servlet:1.3.2
|    +--- com.slack.api:slack-api-model:1.3.2 (*)
|    +--- com.slack.api:slack-api-client:1.3.2 (*)
|    +--- com.slack.api:slack-app-backend:1.3.2 (*)
|    \--- com.slack.api:bolt:1.3.2 (*)
+--- com.slack.api:bolt:1.3.2
|    +--- com.slack.api:slack-api-model:1.3.2
|    +--- com.slack.api:slack-api-client:1.3.2
|    |    +--- com.slack.api:slack-api-model:1.3.2 (*)
|    \--- com.slack.api:slack-app-backend:1.3.2
|         +--- com.slack.api:slack-api-model:1.3.2 (*)
|         +--- com.slack.api:slack-api-client:1.3.2 (*)
+--- com.slack.api:slack-api-client:1.3.2 (*)
+--- com.slack.api:slack-api-model-kotlin-extension:1.3.2
|    +--- com.slack.api:slack-api-model:1.3.2 (*)
+--- com.slack.api:slack-api-client-kotlin-extension:1.3.2
|    +--- com.slack.api:slack-api-model-kotlin-extension:1.3.2 (*)
|    +--- com.slack.api:slack-api-client:1.3.2 (*)
+--- com.slack.api:bolt-servlet:1.3.2
|    +--- com.slack.api:slack-api-model:1.3.2 (*)
|    +--- com.slack.api:slack-api-client:1.3.2 (*)
|    +--- com.slack.api:slack-app-backend:1.3.2 (*)
|    \--- com.slack.api:bolt:1.3.2 (*)
```

#### Java Runtime version

```
openjdk version "14.0.2" 2020-07-14
OpenJDK Runtime Environment (build 14.0.2+12-46)
OpenJDK 64-Bit Server VM (build 14.0.2+12-46, mixed mode, sharing)
```

#### OS info

```
ProductName:	macOS
ProductVersion:	11.0.1
BuildVersion:	20B29
Darwin Kernel Version 20.1.0: Sat Oct 31 00:07:11 PDT 2020; root:xnu-7195.50.7~2/RELEASE_X86_64
```

#### Steps to reproduce:

1. Create an action block with a select menu and an action block with a button
2. Pick an item from the select menu
3. Press the button

#### Expected result:

I expected to be able to access the state property on my handler:

```
app.blockAction("save") { req, ctx ->
    // req.payload.state.values should be defined
    ....
}
```
I believe this is what is described here https://api.slack.com/reference/interaction-payloads/block-actions#fields under `state.values`

Thanks for looking into this. Let me know if you'd like me to provide more information or if you have any questions.

#### Actual result:

The state is null:
```
app.blockAction("save") { req, ctx ->
    // req.payload.state.values is not defined
    ....
}
```

### Requirements

Please make sure if this topic is specific to this SDK. For general questions/issues about Slack API platform or its server-side, could you submit questions at https://my.slack.com/help/requests/new instead. :bow:

Please read the [Contributing guidelines](https://github.com/slackapi/java-slack-sdk/blob/main/.github/contributing.md) and [Code of Conduct](https://slackhq.github.io/code-of-conduct) before creating this issue or pull request. By submitting, you are agreeing to the those rules.

**Comment by seratch** [comment]
You can access the values via `req.payload.view.state.values` for sure. Can you try the following code snippet out?

### build.gradle

```gradle
plugins {
  id("org.jetbrains.kotlin.jvm") version "1.4.10"
  id("application")
}
repositories {
  mavenCentral()
}
dependencies {
  implementation(platform("org.jetbrains.kotlin:kotlin-bom"))
  implementation("org.jetbrains.kotlin:kotlin-stdlib-jdk8")
  implementation("com.slack.api:bolt-jetty:1.3.2")
  implementation("com.slack.api:slack-api-model-kotlin-extension:1.3.2")
  implementation("com.slack.api:slack-api-client-kotlin-extension:1.3.2")
  implementation("org.slf4j:slf4j-simple:1.7.30") // or logback-classic
}
application {
  mainClassName = "hello.MyAppKt" // add "Kt" suffix for main function source file
}
```

### src/main/kotlin/hello/MyApp.kt

```kotlin
package hello

import com.slack.api.app_backend.interactive_components.response.Option
import com.slack.api.bolt.App
import com.slack.api.bolt.jetty.SlackAppServer
import com.slack.api.model.block.composition.BlockCompositions.plainText
import com.slack.api.model.block.composition.PlainTextObject
import com.slack.api.model.kotlin_extension.view.blocks
import com.slack.api.model.view.Views.*
import java.net.URLDecoder

// export SLACK_SIGNING_SECRET=
// export SLACK_BOT_TOKEN=
// gradle run
fun main() {
  System.setProperty("org.slf4j.simpleLogger.log.com.slack.api", "debug")

  val app = App()

  app.use { req, _, chain ->
    if (req.requestBodyAsString.startsWith("payload=")) {
      val body = req.requestBodyAsString.split("payload=")[1]
      val json = URLDecoder.decode(body, "UTF-8")
      req.context.logger.info(json)
    }
    chain.next(req)
  }

  app.command("/open-modal") { _, ctx ->
    val view = view { v ->
      v.type("modal")
        .callbackId("view-id")
        .title(viewTitle { it.type("plain_text").text("Title") })
        .submit(viewSubmit { it.type("plain_text").text("Submit") })
        .blocks {
          actions {
            blockId("block_1")
            elements {
              externalSelect {
                actionId("select")
                minQueryLength(0)
              }
              button {
                actionId("save")
                text("Save")
                value("1")
              }
            }
          }
          input {
            blockId("block_2")
            label("Plain Text Input")
            element {
              plainTextInput {
                actionId("input")
              }
            }
          }
        }
    }
    val apiResult = ctx.client().viewsOpen { it.triggerId(ctx.triggerId).view(view) }
    if (apiResult.isOk) {
      ctx.ack()
    } else {
      ctx.ackWithJson(apiResult)
    }
  }

  val allOptions = listOf(
    Option(plainText("Schedule"), "schedule"),
    Option(plainText("Budget"), "budget"),
    Option(plainText("Assignment"), "assignment")
  )
  app.blockSuggestion("select") { req, ctx ->
    ctx.logger.info("view.state.values: {}", req.payload.view.state.values)
    val keyword = req.payload.value
    if (keyword == null || keyword.isEmpty()) {
      ctx.ack { it.options(allOptions) }
    } else {
      val options = allOptions.filter { (it.text as PlainTextObject).text.contains(keyword) }
      ctx.ack { it.options(options) }
    }
  }

  app.blockAction("save") { req, ctx ->
    ctx.logger.info("view.state.values: {}", req.payload.view.state.values)
    ctx.ack()
  }
  app.blockAction("select") { req, ctx ->
    ctx.logger.info("view.state.values: {}", req.payload.view.state.values)
    ctx.ack()
  }
  app.viewSubmission("view-id") { req, ctx ->
    ctx.logger.info("view.state.values: {}", req.payload.view.state.values)
    ctx.ack()
  }

  val server = SlackAppServer(app)
  server.start()
}
```

**Comment by lfbrandao** [comment]
Thanks @seratch.

The example you sent me works well, but my use case is a little different. I'm not creating a new view - I'm sending a message to the user with interactive components. When I do that, I can't read the state anymore, but I still see a state property on the JSON sent from Slack:

```
{
	"type": "block_actions",
	...
	"state": {
		"values": {
			"project-selector": {
				"select": {
					"type": "external_select",
					"selected_option": {
						"text": {
							"type": "plain_text",
							"text": "Budget",
							"emoji": true
						},
						"value": "budget"
					}
				}
			}
		}
	},
	"actions": [{
		"type": "external_select",
		"action_id": "select",
		"block_id": "project-selector",
		"selected_option": {
			"text": {
				"type": "plain_text",
				"text": "Budget",
				"emoji": true
			},
			"value": "budget"
		},
		"action_ts": "..."
	}]
}
```

I updated your example with a new slash command that is similar to what I'm doing and pushed it here  https://github.com/lfbrandao/slack-state-debug/blob/master/src/main/kotlin/hello/MyApp.kt#L80

Perhaps there's a state property missing on the BlockActionPayload that would contain this information?

**Comment by seratch** [comment]
@lfbrandao In your first message, you said:

>I have a view with a few External Options Selects and a button

so, I assumed you are talking about modal views or home tabs.

>The example you sent me works well, but my use case is a little different. I'm not creating a new view - I'm sending a message to the user with interactive components.

But now I understand you mentioned the use case where an app uses the components in a message. And you're right. The top-level `state` in `block_actions` payload is missing in this SDK's `BlockActionPayload` class.

I will fix this issue and will promptly release a new patch version. Thanks for reporting this and being patient in this conversation.

**Comment by seratch** [comment]
This issue is resolved in version 1.3.3

**Comment by lfbrandao** [comment]
It's working! Thanks so much for quick turnaround @seratch
